### ISSUE TITLE
GetTypeInfo returns NotNull for places where the variable isn't yet assigned

### ISSUE THREAD

**Issue by jasonmalinowski** [dotnet/roslyn#36052]
Imagine this case:

```
string? s;
Foo(s);
```

When we call GetTypeInfo passing in the argument `s`, we get the flow state being NotNull. Although the question is somewhat meaningless in the case of an unassigned variable, it means GenerateMember invoked on Foo will generate the parameter with `string`. I'd propose that GetTypeInfo just returned the annotation in the case the variable is known to not be assigned.

Chatting with @333fred he observed the nullable analysis doesn't have access to definite analysis so that's why it's doing that. However we also observed that there's no way for the IDE to know _	either_ what is going on here short of us calling GetDiagnostics() and manually looking for diagnostics about use of uninitialized variables.

**Comment by jasonmalinowski** [comment]
This behavior for the IDE of course assumes that we're using flow state instead of original annotation to decide which nullability to generate with. I assumed that made sense but there's other cases where it doesn't make sense. We'll discuss what we think the behavior should be further as a design question.

**Comment by gafter** [comment]
The behavior of the API is correct (by design).  The variable will not be null at the call site on line 2 (in part because the code will not compile).

**Comment by jasonmalinowski** [comment]
@gafter Why force the user to do that? How does that help them? The IDE takes the usual tact that we should operate gracefully in incomplete code, and as much as possible shouldn't have to force users to 'correct' code before performing refactorings.

I'm adding Need Design Review back on for the IDE, since we're going to discuss further what we think the IDE experience should be. There was also a practical question of whether we should always be using the declared annotation rather than the flow state. There's arguments both ways and I will admit to having picked one while implementing this without thinking about the other way enough. My concern is if we do decide that we should use flow state _and_ we want this to generate with a nullable parameter, either this API will have to change or we'll need a new API to know that a variable is not definitely assigned at a point so we know to ignore flow state.

**Comment by gafter** [comment]
@jasonmalinowski Why should the user be forced to assign the variable before use???

Your assumption appears to be that the user intends the API to work with a nullable value, and I'm not sure why you conclude that.  Yes, `s` is capable of holding a `null`, but that doesn't mean it will hold `null` when that line is executed.  We just don't know, because the assignment to `s` is missing.

**Comment by jasonmalinowski** [comment]
I guess in absence of further information, the declaration is the only information you've got. Why ignore it?

**Comment by CyrusNajmabadi** [comment]
I agree with jason here. In the absence of other information, it seems totally reasonable to say "you declared it as `string?` so it seems reasonable to default to that".  Virtually every IDE feature i can think of will end up needing to do that AFAICT, so it seems pretty fair to just default to that.

**Comment by 333fred** [comment]
IIRC @gafter's reasoning here was that declaring it to be `NotNull` to start would be the best way to suppress cascading errors. For example:
```C#
string? s;
s.ToString(); // Already going to get an error about use-before-assignment, if it's considered not null then you won't also get a nullability warning.
```

That being, I was and still am more in favor of using the variable's declared type here.

**Comment by jasonmalinowski** [comment]
Chatting with the IDE team we indeed agree that yes, in the case of a variable that's not assigned, we should be falling back with the declared type. We too thought about various features and could think of many features that need this and no features that _want_ this behavior, so we're definitely on board with just changing the behavior of the API -- if we had to stitch this together from multiple APIs we'd probably just being do that everywhere in the IDE to the point we'd ban GetTypeInfo() entirely, and that's not a good sign for the health of the API.

I'm removing "resolution: by design" and "needs design review", and at this point this bug is OK to be taken by the compiler at any point. We'll dupe other bugs against this one.
